### PR TITLE
fix: fix storing of multiclass predictions in CLI

### PR DESCRIPTION
**Pull request by m-volk**
## Description
I used the CLI command to do predictions for a multiclass classification, and noticed the resulting file is broken. The problem is that the exponent of scientific numbers is cut off. The underlying problem seems to be the behavior of `np.apply_along_axis` that infers the type from the first element. [Here](https://stackoverflow.com/questions/55984973/numpy-apply-along-axis-truncates-strings-because-it-infers-wrong-dtype-u1/55985137) is a related discussion. The problem exists in latest main (commit dce2ffd1f61f8069dae40f88e725f17e326488a7).

### The problem in a nutshell
See for example [here](https://github.com/chemprop/chemprop/blob/main/chemprop/cli/predict.py#L379).
```
import torch
import numpy as np

print("Current behavior:")
test_preds = torch.tensor([[[3.3954e-01, 6.3595e-01, 2.4509e-02]],
        [[9.8928e-01, 1.0667e-02, 4.8996e-05]]])

formatted_probability_strings = np.apply_along_axis(
    lambda x: ",".join(map(str, x)), 2, test_preds.numpy()
)
print(formatted_probability_strings)

> Current behavior:
> [['0.33954,0.63595,0.024509']
>  ['0.98928,0.010667,4.8996e']]
```
The problem is the number in the lower right. The exponent is missing.

### Code example
Run following code with the attached checkpoint and test data.
```
chemprop predict \
    --test-path test_data.csv \
    -o my_prediction.csv \
    -s smiles \
    --model-paths best-epoch=20-val_loss=0.52.ckpt
```

[example.zip](https://github.com/user-attachments/files/20613515/example.zip)

## Bugfix / Desired workflow

There are many potential solutions to solve this. The smallest code change that I can think of, is to enforce the scientific notation for all numbers.
```
import torch
import numpy as np

print("Proposed behavior:")
def to_str(z):
    return f"{z:.6e}"

formatted_probability_strings = np.apply_along_axis(
    lambda x: ",".join([to_str(y) for y in x]), 2, test_preds.numpy()
)
print(formatted_probability_strings)

> Proposed behavior:
> [['3.395400e-01,6.359500e-01,2.450900e-02']
>  ['9.892800e-01,1.066700e-02,4.899600e-05']]
```

## Questions
The downside of my proposal is that some people might prefer the fixed-point representation, when possible.

## Checklist
- [X] linted with flake8?
- [ ] (if appropriate) unit tests added?

### CODE REVIEW

**Comment by m-volk** [comment]
Hi @JacksonBurns! I think you are right, `apply_along_axis` is kind of an overkill here. The string conversion is not the bottle neck of this tool, so there is no need to make its execution efficient with this numpy command. On the other hand, I think that `apply_along_axis` is a very elegant way to write the command. My feeling is that your code snippet doesn't exactly reproduce the original behavior. I had run the tests in `tests/cli/` and recorded the input and output to the `apply_along_axis` command, and experimented a bit.
What do you think about factoring the `apply_along_axis` calls out into `utils.py`? Then we can have some unittests there. With the unittests it should be easier to refactor the function into a style that you prefer.

**Review by JacksonBurns — APPROVED** [review]

**Review by akshatzalte — APPROVED** [review]
Overall this is good to go, approved. I have made very minor changes in `test_cli_utils.py`. Will merge it once all tests pass.
